improve doc on how to understand dynamo (#106860)
pytorch/pytorch: docs/source/torch.compiler_deepdive.rst

function to be executed in a new Python frame which recursively
triggers TorchDynamo to restart its capture once execution reaches that
point for the first time.

> 
> How to retrieve artifacts generated by Dynamo?
> ----------------------------------------------
> 
> If you want to inspect the artifacts generated by Dynamo, here is what you can do:
> 
> First, we have an API `torch._dynamo.eval_frame._debug_get_cache_entry_list` to retrieve compiled code and guards out of a compiled function. A compiled function can have several cache entries, and each cache entry consists a lambda function to keep the guards, and a codeobject to keep the code to be executed if the guarding conditions are satisfied.
> 
> .. code-block:: python
> 
>    from torch._dynamo.eval_frame import _debug_get_cache_entry_list
>    cache_entries = _debug_get_cache_entry_list(toy_example._torchdynamo_orig_callable.__code__)
>    guard, code = cache_entries[0]
>    # the guard takes a input frame, and tells whether a re-compilation should be triggered.
>    import inspect
>    print(inspect.getfullargspec(guard))
>    # if you know python bytecode, you can understand the following code.
>    import dis
>    dis.dis(guard)
>    dis.dis(code)
> 
> Second, the compiled function for subgraphs like `__compiled_fn_0` has already been released into your python scope now! You can find all the functions in the output of `dis.dis(code)`.
> 
> .. code-block:: python
> 
>    # Try this out! You didn't define the function, but Dynamo implicitly defines for you!
>    __compiled_fn_0
> 
> With the above utilities and explanations, you can now manually inspect the compiled functions and guards as you like, so long as you can read the python bytecode.
> 
> Manually decompilation of the bytecode looks like the following:
> 
> .. code-block:: python
> 
>    def compiled_example(a, b):
>        x, _ = __compiled_fn_0(a, b)
>        if not _:
>            return __resume_at_38_2(b, x)
>         return __resume_at_30_1(b, x)
> 
>    def __resume_at_38_2(b, x):
>        return x * b
> 
>    def __resume_at_30_1(b, x):
>        b = b * -1
>        return x * b
> 
> The `__compiled_fn_0` is more complicated, though. It will invoke the guards, and check the guarding conditions, trigger re-compilation if needed.

Reviewer: ```suggestion

How to inspect artifacts generated by TorchDynamo?
----------------------------------------------

To inspect the artifacts generated by TorchDynamo, there is an API `torch._dynamo.eval_frame._debug_get_cache_entry_list` that retrieves compiled code and guards out of a function's `__code__` object. A compiled function can have several cache entries, and each cache entry consists a generated function to check guards, and a `types.CodeType` object to keep the code to be executed if the guarding conditions are satisfied.

.. code-block:: python

   from torch._dynamo.eval_frame import _debug_get_cache_entry_list
   cache_entries = _debug_get_cache_entry_list(toy_example._torchdynamo_orig_callable.__code__)
   guard, code = cache_entries[0]
   # the guard takes a input frame, and tells whether a re-compilation should be triggered.
   import inspect
   print(inspect.getfullargspec(guard))
   # if you know python bytecode, you can understand the following code.
   import dis
   dis.dis(guard)
   dis.dis(code)

The compiled bytecode, printed by `dis.dis(code)`, will call the result of the backend compiler function which is stored in a global variable such as `__compiled_fn_0` in the module containing the original function.

The generated bytecodes are roughly equivilent to the following Python (converted manually for illustration purposes).

.. code-block:: python

   def compiled_example(a, b):
       x, _ = __compiled_fn_0(a, b)
       if not _:
           return __resume_at_38_2(b, x)
        return __resume_at_30_1(b, x)

   def __resume_at_38_2(b, x):
       return x * b

   def __resume_at_30_1(b, x):
       b = b * -1
       return x * b
```

> The `__compiled_fn_0` is more complicated, though. It will invoke the guards, and check the guarding conditions, trigger re-compilation if needed.

Reviewer: > The `__compiled_fn_0` is more complicated, though. It will invoke the guards, and check the guarding conditions, trigger re-compilation if needed.

This isn't exactly right.  `__compiled_fn_0` is whatever you passed as `backend=` to `torch.compile` returns.  In the case of `backend=lambda graph, _: graph`, it will actually be a pretty readable FX graph of ops.  With other backends it will be generated code.

Author: Then where is the logic of "invoke the guards, and check the guarding conditions, trigger re-compilation if needed"?

Reviewer: Guards are checked here:
https://github.com/pytorch/pytorch/blob/ddd2f682b974fa274771965266c2bc0786f1e747/torch/csrc/dynamo/eval_frame.c#L421

If that function returns Py_None it recompiles.

Author: This is my output of `dis.dis(__compiled_fn_0)`. I don't see the output of the return from backend. Only some torchdynamo stuff.

```
315           0 LOAD_GLOBAL              0 (isinstance)
              2 LOAD_DEREF               4 (self)
              4 LOAD_GLOBAL              1 (DisableContext)
              6 CALL_FUNCTION            2

314           8 POP_JUMP_IF_TRUE        25 (to 50)

316          10 LOAD_GLOBAL              2 (torch)
             12 LOAD_ATTR                3 (fx)
             14 LOAD_ATTR                4 (_symbolic_trace)
             16 LOAD_METHOD              5 (is_fx_tracing)
             18 CALL_METHOD              0

314          20 POP_JUMP_IF_FALSE       25 (to 50)

318          22 LOAD_GLOBAL              6 (config)
             24 LOAD_ATTR                7 (error_on_nested_fx_trace)
             26 POP_JUMP_IF_FALSE       18 (to 36)

319          28 LOAD_GLOBAL              8 (RuntimeError)

320          30 LOAD_CONST               1 ('Detected that you are using FX to symbolically trace a dynamo-optimized function. This is not supported at the moment.')

319          32 CALL_FUNCTION            1
             34 RAISE_VARARGS            1

324     >>   36 LOAD_DEREF               2 (fn)
             38 LOAD_FAST                0 (args)
             40 BUILD_MAP                0
             42 LOAD_FAST                1 (kwargs)
             44 DICT_MERGE               1
             46 CALL_FUNCTION_EX         1
             48 RETURN_VALUE

326     >>   50 LOAD_DEREF               3 (on_enter)
             52 CALL_FUNCTION            0
             54 POP_TOP

327          56 LOAD_GLOBAL              9 (set_eval_frame)
             58 LOAD_DEREF               1 (callback)
             60 CALL_FUNCTION            1
             62 STORE_FAST               2 (prior)

328          64 LOAD_DEREF               0 (backend_ctx_ctor)
             66 CALL_FUNCTION            0
             68 STORE_FAST               3 (backend_ctx)

329          70 LOAD_FAST                3 (backend_ctx)
             72 LOAD_METHOD             10 (__enter__)
             74 CALL_METHOD              0
             76 POP_TOP

330          78 LOAD_GLOBAL             11 (enable_dynamic)
             80 LOAD_DEREF               4 (self)
             82 LOAD_ATTR               12 (dynamic)
             84 LOAD_DEREF               4 (self)
             86 LOAD_ATTR               13 (export)
             88 CALL_FUNCTION            2
             90 STORE_FAST               4 (dynamic_ctx)

331          92 LOAD_FAST                4 (dynamic_ctx)
             94 LOAD_METHOD             10 (__enter__)
             96 CALL_METHOD              0
             98 POP_TOP

332         100 SETUP_FINALLY           26 (to 154)

333         102 LOAD_DEREF               2 (fn)
            104 LOAD_FAST                0 (args)
            106 BUILD_MAP                0
            108 LOAD_FAST                1 (kwargs)
            110 DICT_MERGE               1
            112 CALL_FUNCTION_EX         1
            114 POP_BLOCK

335         116 LOAD_GLOBAL              9 (set_eval_frame)
            118 LOAD_FAST                2 (prior)
            120 CALL_FUNCTION            1
            122 POP_TOP

336         124 LOAD_FAST                4 (dynamic_ctx)
            126 LOAD_METHOD             14 (__exit__)
            128 LOAD_CONST               0 (None)
            130 LOAD_CONST               0 (None)
            132 LOAD_CONST               0 (None)
            134 CALL_METHOD              3
            136 POP_TOP

337         138 LOAD_FAST                3 (backend_ctx)
            140 LOAD_METHOD             14 (__exit__)
            142 LOAD_CONST               0 (None)
            144 LOAD_CONST               0 (None)
            146 LOAD_CONST               0 (None)
            148 CALL_METHOD              3
            150 POP_TOP
            152 RETURN_VALUE

335     >>  154 LOAD_GLOBAL              9 (set_eval_frame)
            156 LOAD_FAST                2 (prior)
            158 CALL_FUNCTION            1
            160 POP_TOP

336         162 LOAD_FAST                4 (dynamic_ctx)
            164 LOAD_METHOD             14 (__exit__)
            166 LOAD_CONST               0 (None)
            168 LOAD_CONST               0 (None)
            170 LOAD_CONST               0 (None)
            172 CALL_METHOD              3
            174 POP_TOP

337         176 LOAD_FAST                3 (backend_ctx)
            178 LOAD_METHOD             14 (__exit__)
            180 LOAD_CONST               0 (None)
            182 LOAD_CONST               0 (None)
            184 LOAD_CONST               0 (None)
            186 CALL_METHOD              3
            188 POP_TOP
            190 RERAISE                  0
```

Author: Okay, I got it. `__compiled_fn_0._torchdynamo_orig_callable` is what returned from the backend :)

function to be executed in a new Python frame which recursively
triggers TorchDynamo to restart its capture once execution reaches that
point for the first time.

How to inspect artifacts generated by TorchDynamo?
----------------------------------------------

To inspect the artifacts generated by TorchDynamo, there is an API `torch._dynamo.eval_frame._debug_get_cache_entry_list` that retrieves compiled code and guards out of a function's `__code__` object. A compiled function can have several cache entries, and each cache entry consists a generated function to check guards, and a `types.CodeType` object to keep the code to be executed if the guarding conditions are satisfied.

.. code-block:: python

>    from torch._dynamo.eval_frame import _debug_get_cache_entry_list

Reviewer 2: so what is interesting about the printed output here? or like what does it add on top of the printed assembly code from the section above

Author: The section above prints all the compiled code, which is intimidating and hard to understand for large functions (like a whole neural network). This API lets user to extract compiled code for **each individual function**, which is more clear and gives the user more control.

   cache_entries = _debug_get_cache_entry_list(toy_example._torchdynamo_orig_callable.__code__)
   guard, code = cache_entries[0]
   # the guard takes an input frame, and tells whether a re-compilation should be triggered.
   import inspect
   print(inspect.getfullargspec(guard))
   # if you know python bytecode, you can understand the following code.
   import dis
   dis.dis(guard)
   dis.dis(code)

The compiled bytecode, printed by `dis.dis(code)`, will call the result of the backend compiler function which is stored inside a global variable such as `__compiled_fn_0` in the module containing the original function.

> The generated bytecodes are roughly equivalent to the following Python (converted manually for illustration purposes).

Reviewer 2: a bit bummed that it seems python doesnt have a library to go from disassembly to these python functions but otherwise this fine

Author: Yep, python lacks a decent `decompiler` to go from disassembly to python code.

.. code-block:: python

>    def compiled_example(a, b):

Reviewer 2: add a comment showing the graph break, that's an important part

Author: Fixed in [12e515a](https://github.com/pytorch/pytorch/pull/106860/commits/12e515ab8d72e09daa1155c45d75320ef4453a11).

       # behind the scene, pytorch C code checks the guarding condition
       # if all guard fails, trigger re-compile
       # else, run the compiled code
       # after some setup work, the code finally looks like the following
       x, _ = __compiled_fn_0._torchdynamo_orig_callable(a, b)

>        if not _:

Reviewer 2: what's `_` could you name the variable?

Author: It's a temp variable to hold the result of `b.sum() < 0`. Fixed in [e3fc8a3](https://github.com/pytorch/pytorch/pull/106860/commits/e3fc8a3f426f9075404d9bd843d8edfc5c7b40d7).

           return __resume_at_38_2(b, x)
        return __resume_at_30_1(b, x)

   def __resume_at_38_2(b, x):
       return x * b

   def __resume_at_30_1(b, x):
       b = b * -1
       return x * b

   def fn(a, b):
       x = a / (torch.abs(a) + 1)
       lt = b.sum() < 0
       return x, lt

   __compiled_fn_0._torchdynamo_orig_callable = fn

> Note that we pass a simple `my_compiler` function as the backend compiler, therefore the subgraph code `__resume_at_38_2`, `__resume_at_30_1`, and `__compiled_fn_0._torchdynamo_orig_callable` remain python code. However, if we use other backends like the built-in `torchinductor`, the subgraph code will be compiled CUDA kernels for GPU or C++ code for CPU.

Reviewer 2: ```suggestion
Note that we pass a simple `my_compiler` function as the backend compiler, therefore the subgraph code `__resume_at_38_2`, `__resume_at_30_1`, and `__compiled_fn_0._torchdynamo_orig_callable` remain python code. However, if we use other backends like the built-in `inductor`, the subgraph code will be compiled CUDA kernels for GPU or C++ code for CPU.
```

   dis.dis(guard)
   dis.dis(code)

The compiled bytecode, printed by `dis.dis(code)`, will call the result of the backend compiler function which is stored in a global variable such as `__compiled_fn_0` in the module containing the original function.

> The generated bytecodes are roughly equivilent to the following Python (converted manually for illustration purposes).

Reviewer: ```suggestion
The generated bytecodes are roughly equivalent to the following Python (converted manually for illustration purposes).
```
